src/Lykke.Service.EasyBuy.DomainServices/PricesService.cs

```diff
@@ -0,0 +1,143 @@
+        }
+        
+        public async Task<PriceSnapshot> CreateSnapshotAsync(string assetPair, OrderType type, decimal quotingVolume, DateTime validFrom)
+        {
+            var instrument = await _instrumentsAccessService.GetByAssetPairIdAsync(assetPair);
+
+            if (instrument == null || instrument.State != InstrumentState.Active)
+            {
+                throw new OperationFailedException($"No active instrument {assetPair} was found.");
+            }
+
+            var orderBook = _orderBookService.GetByAssetPairId(instrument.Exchange, assetPair);
+
+            if(type == OrderType.None)
+                throw new OperationFailedException($"Invalid order type {OrderType.None}");
+
+            var pair = await _assetsService.TryGetAssetPairAsync(assetPair);
+            
+            if(pair == null)
+                throw new OperationFailedException($"Pair {assetPair} not found.");
+
+            var baseAsset = await _assetsService.TryGetAssetAsync(pair.BaseAssetId);
+            var quotingAsset = await _assetsService.TryGetAssetAsync(pair.QuotingAssetId);
+            
+            if(baseAsset == null)
+                throw new OperationFailedException($"Base asset {pair.BaseAssetId} not found.");
+            
+            if(quotingAsset == null)
+                throw new OperationFailedException($"Quoting asset {pair.QuotingAssetId} not found.");
+
+            var orders = type == OrderType.Buy
+                ? orderBook.SellLimitOrders.OrderBy(x => x.Price).ToList()
+                : orderBook.BuyLimitOrders.OrderByDescending(x => x.Price).ToList();
+            
+            if (!orders.Any() || orders.Sum(x => x.Volume * x.Price) < quotingVolume)
+            {
+                throw new OperationFailedException("Not enough liquidity.");
+            }
+
+            var markup = instrument.Markup ?? await _settingsService.GetDefaultMarkupAsync();
+            
+            var ordersInUsd = orders.Select(x => x.Volume * (x.Price + markup * (type == OrderType.Sell ? -1 : 1))).ToList();
+                
+            var volumes = new List<decimal>();
+            var remainingVolume = quotingVolume;
+
+            for (var i = 0; i < ordersInUsd.Count; i++)
+            {
+                var orderVolumeInUsd = ordersInUsd[i];
+                    
+                if (remainingVolume == 0m || remainingVolume <= (decimal) pair.MinVolume)
+                    break;
+
+                if (orderVolumeInUsd <= remainingVolume)
+                {
+                    volumes.Add(orders[i].Volume);
+
+                    remainingVolume -= orderVolumeInUsd;
+                }
+                else
+                {
+                    volumes.Add(remainingVolume / (orders[i].Price + markup * (type == OrderType.Sell ? -1 : 1)));
+
+                    break;
+                }
+            }
+
+            var volumeBase = volumes.Sum().TruncateDecimalPlaces(baseAsset.Accuracy, type == OrderType.Sell);
```

Comment:
The volume rounded be mathematical rules. Price rounded using TruncateDecimalPlaces method.